src/WorkflowFunction.ts

```diff
@@ -0,0 +1,214 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import {
+  WebClient,
+  FunctionsCompleteErrorResponse,
+  FunctionsCompleteSuccessResponse,
+} from '@slack/web-api';
+import {
+  Middleware,
+  AllMiddlewareArgs,
+  AnyMiddlewareArgs,
+  SlackEventMiddlewareArgs,
+  Context,
+} from './types';
+import processMiddleware from './middleware/process';
+import { WorkflowFunctionInitializationError } from './errors';
+
+/** Interfaces */
+
+export interface FunctionCompleteArguments {
+  outputs?: {
+    [key: string]: any;
+  };
+}
+
+export interface FunctionFailArguments {
+  error: string;
+}
+
+export interface FunctionCompleteFn {
+  (params?: FunctionCompleteArguments): Promise<FunctionsCompleteSuccessResponse>;
+}
+
+export interface FunctionFailFn {
+  (params: FunctionFailArguments): Promise<FunctionsCompleteErrorResponse>;
+}
+
+export interface WorkflowFunctionExecuteMiddlewareArgs extends SlackEventMiddlewareArgs<'function_executed'> {
+  complete: FunctionCompleteFn;
+  fail: FunctionFailFn;
+}
+
+/** Types */
+
+export type SlackWorkflowFunctionMiddlewareArgs = WorkflowFunctionExecuteMiddlewareArgs;
+
+export type WorkflowFunctionExecuteMiddleware = Middleware<WorkflowFunctionExecuteMiddlewareArgs>;
+
+export type WorkflowFunctionMiddleware = WorkflowFunctionExecuteMiddleware[];
+
+export type AllWorkflowFunctionMiddlewareArgs
+  <T extends SlackWorkflowFunctionMiddlewareArgs = SlackWorkflowFunctionMiddlewareArgs> = T & AllMiddlewareArgs;
+
+/** Constants */
+
+const VALID_PAYLOAD_TYPES = new Set(['function_executed']);
+
+/** Class */
+
+export class WorkflowFunction {
```

Comment:
I can't remember what naming conventions we have to follow surrounding this but I think `CustomFunction` might align with what was the official verbiage (Can't find the document where this was stated anymore)

Reply:
With our recent discussions floating around about naming, I'm actually more in favor of that than what's here. Thanks for suggesting it!

Reply:
Good call Will, let's coalesce around the terms in our public API docs whenever possible, to keep terminology consistent.